File: 1176-design-a-leaderboard/design-a-leaderboard.java
class Leaderboard {
    
    private Map<Integer, Integer> scores;

    public Leaderboard() {
        scores = new HashMap<>();
    }

    public void addScore(int playerId, int score) {
        scores.put(playerId, scores.getOrDefault(playerId, 0) + score);
    }

    public int top(int K) {
        Queue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        for (Map.Entry<Integer, Integer> entry : scores.entrySet()) {
            pq.add(entry.getValue());
        }
        int res = 0;
        for (int i = 0; i < K; i++) {
            res += pq.poll();
        }
        return res;
    }

    public void reset(int playerId) {
        scores.remove(playerId);
    }
}

/**
 * Your Leaderboard object will be instantiated and called as such:
 * Leaderboard obj = new Leaderboard();
 * obj.addScore(playerId,score);
 * int param_2 = obj.top(K);
 * obj.reset(playerId);
 */

File: 1176-design-a-leaderboard/design-a-leaderboard.py
class Leaderboard:

    def __init__(self):
        self.map = collections.defaultdict(int)

    def addScore(self, playerId: int, score: int) -> None:  # log(n)
        self.map[playerId] += score

    def top(self, K: int) -> int:
        max_heap = []
        for p, s in self.map.items():
            heapq.heappush(max_heap, -s)
        topK = 0
        for _ in range(K):
            topK += heapq.heappop(max_heap)
        return -topK

    def reset(self, playerId: int) -> None:     # 1
        self.map.pop(playerId)


# Your Leaderboard object will be instantiated and called as such:
# obj = Leaderboard()
# obj.addScore(playerId,score)
# param_2 = obj.top(K)
# obj.reset(playerId)